File: ios/Classes/YandexPlacemarkController.swift
import YandexMapsMobile

class YandexPlacemarkController:
  NSObject,
  YandexMapObjectController,
  YMKMapObjectTapListener,
  YMKMapObjectDragListener
{
  private let internallyControlled: Bool
  private let parent: YMKMapObject // Workaround https://github.com/yandex/mapkit-ios-demo/issues/100
  public let placemark: YMKPlacemarkMapObject
  private var consumeTapEvents: Bool = false
  public unowned var controller: YandexMapController
  public let id: String

  public required init(
    parent: YMKMapObject,
    params: [String: Any],
    controller: YandexMapController
  ) {
    var placemark: YMKPlacemarkMapObject? = nil
    let point = Utils.pointFromJson(params["point"] as! [String: NSNumber])

    if (parent is YMKClusterizedPlacemarkCollection) {
      placemark = (parent as! YMKClusterizedPlacemarkCollection).addPlacemark(with: point)
    }

    if (parent is YMKMapObjectCollection) {
      placemark = (parent as! YMKMapObjectCollection).addPlacemark(with: point)
    }

    self.parent = parent
    self.placemark = placemark!
    self.id = params["id"] as! String
    self.controller = controller
    self.internallyControlled = false

    super.init()

    placemark!.userData = self.id
    placemark!.addTapListener(with: self)
    placemark!.setDragListenerWith(self)
    update(params)
  }

  public required init(
    parent: YMKMapObject,
    placemark: YMKPlacemarkMapObject,
    params: [String: Any],
    controller: YandexMapController
  ) {
    self.parent = parent
    self.placemark = placemark
    self.id = params["id"] as! String
    self.controller = controller
    self.internallyControlled = true

    super.init()

    placemark.userData = self.id
    placemark.addTapListener(with: self)
    placemark.setDragListenerWith(self)
    update(params)
  }

  public func update(_ params: [String: Any]) {
    if (!internallyControlled) {
      placemark.geometry = Utils.pointFromJson(params["point"] as! [String: NSNumber])
      placemark.isVisible = (params["isVisible"] as! NSNumber).boolValue
    }

    placemark.zIndex = (params["zIndex"] as! NSNumber).floatValue
    placemark.isDraggable = (params["isDraggable"] as! NSNumber).boolValue
    placemark.opacity = (params["opacity"] as! NSNumber).floatValue
    placemark.direction = (params["direction"] as! NSNumber).floatValue

    setIcon(params["icon"] as? [String: Any])

    consumeTapEvents = (params["consumeTapEvents"] as! NSNumber).boolValue
  }

  public func remove() {
    if (internallyControlled) {
      return
    }

    if (parent is YMKClusterizedPlacemarkCollection) {
      (parent as! YMKClusterizedPlacemarkCollection).remove(withPlacemark: placemark)
    }

    if (parent is YMKMapObjectCollection) {
      (parent as! YMKMapObjectCollection).remove(with: placemark)
    }
  }

  func onMapObjectDragStart(with mapObject: YMKMapObject) {
    controller.mapObjectDragStart(id: id)
  }

  func onMapObjectDrag(with mapObject: YMKMapObject, point: YMKPoint) {
    controller.mapObjectDrag(id: id, point: point)
  }

  func onMapObjectDragEnd(with mapObject: YMKMapObject) {
    controller.mapObjectDragEnd(id: id)
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    controller.mapObjectTap(id: id, point: point)

    return consumeTapEvents
  }

  private func setIcon(_ icon: [String: Any]?) {
    if (icon == nil) {
      return
    }

    let iconType = icon!["type"] as! String

    if (iconType == "single") {
      let style = icon!["style"] as! [String: Any]
      let image = style["image"] as! [String: Any]

      placemark.setIconWith(getIconImage(image), style: getIconStyle(style))
    }

    if (iconType == "composite") {
      let compositeIcon = placemark.useCompositeIcon()
      let iconParts = icon!["iconParts"] as! [[String: Any]]

      for iconPart in iconParts {
        let style = iconPart["style"] as! [String: Any]
        let image = style["image"] as! [String: Any]
        let name = iconPart["name"] as! String

        compositeIcon.setIconWithName(name, image: getIconImage(image), style: getIconStyle(style))
      }
    }
  }

  private func getIconImage(_ image: [String: Any]) -> UIImage {
    let type = image["type"] as! String
    let defaultImage = UIImage()

    if (type == "fromAssetImage") {
      let assetName = controller.pluginRegistrar.lookupKey(forAsset: image["assetName"] as! String)

      return UIImage(named: assetName) ?? defaultImage
    }

    if (type == "fromBytes") {
      let imageData = (image["rawImageData"] as! FlutterStandardTypedData).data

      return UIImage(data: imageData) ?? defaultImage
    }

    return defaultImage
  }

  private func getIconStyle(_ style: [String: Any]) -> YMKIconStyle {
    let iconStyle = YMKIconStyle()

    if let tappableArea = style["tappableArea"] as? [String: Any] {
      iconStyle.tappableArea = Utils.rectFromJson(tappableArea)
    }

    iconStyle.anchor = NSValue(cgPoint: Utils.rectPointFromJson(style["anchor"] as! [String: NSNumber]))
    iconStyle.zIndex = (style["zIndex"] as! NSNumber)
    iconStyle.scale = (style["scale"] as! NSNumber)
    iconStyle.visible = (style["isVisible"] as! NSNumber)
    iconStyle.flat = (style["isFlat"] as! NSNumber)
    iconStyle.rotationType = YMKRotationType.init(
      rawValue: (style["rotationType"] as! NSNumber).uintValue
    )!.rawValue as NSNumber

    return iconStyle
  }
}

File: android/src/main/java/com/unact/yandexmapkit/YandexPlacemarkController.java
package com.unact.yandexmapkit;

import android.graphics.BitmapFactory;

import androidx.annotation.NonNull;

import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.map.ClusterizedPlacemarkCollection;
import com.yandex.mapkit.map.CompositeIcon;
import com.yandex.mapkit.map.IconStyle;
import com.yandex.mapkit.map.MapObject;
import com.yandex.mapkit.map.MapObjectCollection;
import com.yandex.mapkit.map.MapObjectDragListener;
import com.yandex.mapkit.map.MapObjectTapListener;
import com.yandex.mapkit.map.PlacemarkMapObject;
import com.yandex.mapkit.map.RotationType;
import com.yandex.runtime.image.ImageProvider;

import java.lang.ref.WeakReference;
import java.util.List;
import java.util.Map;

import io.flutter.FlutterInjector;

public class YandexPlacemarkController
  extends YandexMapObjectController
  implements MapObjectTapListener, MapObjectDragListener
{
  private final boolean internallyControlled;
  private final MapObject parent; // Workaround https://github.com/yandex/mapkit-android-demo/issues/258
  public final PlacemarkMapObject placemark;
  private boolean consumeTapEvents = false;
  private final WeakReference<YandexMapController> controller;
  public final String id;

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public YandexPlacemarkController(
    MapObject parent,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    PlacemarkMapObject placemark = null;
    Point point = Utils.pointFromJson((Map<String, Object>) params.get("point"));

    if (parent instanceof ClusterizedPlacemarkCollection) {
      placemark = ((ClusterizedPlacemarkCollection) parent).addPlacemark(point);
    }

    if (parent instanceof MapObjectCollection) {
       placemark = ((MapObjectCollection) parent).addPlacemark(point);
    }

    this.parent = parent;
    this.placemark = placemark;
    this.id = (String) params.get("id");
    this.controller = controller;
    this.internallyControlled = false;

    placemark.setUserData(id);
    placemark.addTapListener(this);
    placemark.setDragListener(this);
    update(params);
  }

  public YandexPlacemarkController(
    MapObject parent,
    PlacemarkMapObject placemark,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    this.parent = parent;
    this.placemark = placemark;
    this.id = (String) params.get("id");
    this.controller = controller;
    this.internallyControlled = true;

    placemark.setUserData(id);
    placemark.addTapListener(this);
    placemark.setDragListener(this);
    update(params);
  }
  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void update(Map<String, Object> params) {
    if (!internallyControlled) {
      placemark.setGeometry(Utils.pointFromJson((Map<String, Object>) params.get("point")));
    }

    placemark.setZIndex(((Double) params.get("zIndex")).floatValue());
    placemark.setVisible((Boolean) params.get("isVisible"));
    placemark.setDraggable((Boolean) params.get("isDraggable"));
    placemark.setOpacity(((Double) params.get("opacity")).floatValue());
    placemark.setDirection(((Double) params.get("direction")).floatValue());

    setIcon(((Map<String, Object>) params.get("icon")));

    consumeTapEvents = (Boolean) params.get("consumeTapEvents");
  }

  public void remove() {
    if (internallyControlled) {
      return;
    }

    if (parent instanceof ClusterizedPlacemarkCollection) {
      ((ClusterizedPlacemarkCollection) parent).remove(placemark);
    }

    if (parent instanceof MapObjectCollection) {
      ((MapObjectCollection) parent).remove(placemark);
    }
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  private void setIcon(Map<String, Object> icon) {
    if (icon == null) {
      return;
    }

    String iconType = ((String) icon.get("type"));

    if (iconType.equals("single")) {
      Map<String, Object> style = ((Map<String, Object>) icon.get("style"));
      Map<String, Object> image = ((Map<String, Object>) style.get("image"));

      placemark.setIcon(getIconImage(image), getIconStyle(style));
    }

    if (iconType.equals("composite")) {
      CompositeIcon compositeIcon = placemark.useCompositeIcon();
      List<Map<String, Object>> iconParts = ((List<Map<String, Object>>) icon.get("iconParts"));

      for (Map<String, Object> iconPart: iconParts) {
        Map<String, Object> style = ((Map<String, Object>) iconPart.get("style"));
        Map<String, Object> image = ((Map<String, Object>) style.get("image"));
        String name = (String) iconPart.get("name");

        compositeIcon.setIcon(name, getIconImage(image), getIconStyle(style));
      }
    }
  }

  @SuppressWarnings({"ConstantConditions"})
  private ImageProvider getIconImage(Map<String, Object> image) {
    String type = (String) image.get("type");

    if (type.equals("fromAssetImage")) {
      return ImageProvider.fromAsset(
        controller.get().context,
        FlutterInjector.instance().flutterLoader().getLookupKeyForAsset((String) image.get("assetName"))
      );
    }
    
    if (type.equals("fromBytes")) {
      byte[] rawImageData = (byte[]) image.get("rawImageData");

      return ImageProvider.fromBitmap(BitmapFactory.decodeByteArray(rawImageData, 0, rawImageData.length));
    }

    return null;
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  private IconStyle getIconStyle(Map<String, Object> style) {
    IconStyle iconStyle = new IconStyle();

    if (style.get("tappableArea") != null) {
      iconStyle.setTappableArea(Utils.rectFromJson((Map<String, Object>) style.get("tappableArea")));
    }

    iconStyle.setAnchor(Utils.rectPointFromJson((Map<String, Double>) style.get("anchor")));
    iconStyle.setZIndex(((Double) style.get("zIndex")).floatValue());
    iconStyle.setScale(((Double) style.get("scale")).floatValue());
    iconStyle.setVisible((Boolean) style.get("isVisible"));
    iconStyle.setFlat((Boolean) style.get("isFlat"));
    iconStyle.setRotationType(RotationType.values()[(Integer) style.get("rotationType")]);
    
    return iconStyle;
  }

  @Override
  public void onMapObjectDragStart(@NonNull MapObject mapObject) {
    controller.get().mapObjectDragStart(id);
  }

  @Override
  public void onMapObjectDrag(@NonNull MapObject mapObject, @NonNull Point point) {
    controller.get().mapObjectDrag(id, point);
  }

  @Override
  public void onMapObjectDragEnd(@NonNull MapObject mapObject) {
    controller.get().mapObjectDragEnd(id);
  }

  @Override
  public boolean onMapObjectTap(@NonNull MapObject mapObject, @NonNull Point point) {
    controller.get().mapObjectTap(id, point);

    return consumeTapEvents;
  }
}
